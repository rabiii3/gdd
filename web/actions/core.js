export const START_REQUEST = 'core/START_REQUEST';
export const END_REQUEST = 'core/END_REQUEST';
export const REQUEST_ERROR = 'core/REQUEST_ERROR';
export const ACK_ERROR = 'core/ACK_ERROR';
export const COMM_ERROR = 'commError';
export const AUTH_ERROR = 'authError';

export const startRequest = () => ({ type: START_REQUEST });
export const endRequest = () => ({ type: END_REQUEST });

export const requestError = ({ method, type, statusCode }) => ({
  type: REQUEST_ERROR,
  error: { method, type, statusCode },
});

export const ackError = () => ({ type: ACK_ERROR });
