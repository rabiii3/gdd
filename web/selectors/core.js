import { pathEq, prop } from 'ramda';
import { createSelector } from 'reselect';

export const root = prop('core');
export const hasSignInError = createSelector(root, pathEq(['error', 'method'], 'auth:signIn'));
export const hasRegisterError = createSelector(root, pathEq(['error', 'method'], 'auth:register'));
export const hasRegisterInError = createSelector(root, pathEq(['error', 'method'], 'auth:registerIn'));
export const getError = createSelector(root, prop('error'));
