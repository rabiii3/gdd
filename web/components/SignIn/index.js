import React from 'react';
import Button from '@material-ui/core/Button';
import { compose } from 'ramda';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { login } from '../../actions/auth';

const SignIn = ({ actions }) => (
  <div>
    <Button
      onClick={() => {
        actions.login({
          _id: '1',
          firstname: 'slim',
          lastname: 'chkir',
          email: 'slim.tunisie@gmail.com',
          roles: ['Admin', 'User'],
        });
      }}
    >
      SignIn as Admin
    </Button>
    <Button
      onClick={() => {
        actions.login({
          _id: '2',
          firstname: 'kadda',
          lastname: 'tarek',
          email: ' kadda.tarek@gmail.com',
          roles: ['user'],
        });
      }}
    >
      SignIn as user
    </Button>
  </div>
);

SignIn.propTypes = {
  actions: PropTypes.object.isRequired,
};
const actions = { login };
const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions, dispatch),
});
const enhance = compose(
  connect(
    null,
    mapDispatchToProps,
  ),
  withRouter,
);
export default enhance(SignIn);
