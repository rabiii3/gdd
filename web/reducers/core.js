import { pathOr } from 'ramda';
import { AUTH_ERROR, COMM_ERROR, START_REQUEST, END_REQUEST, REQUEST_ERROR, ACK_ERROR } from '../actions/core';

const getErrorType = ({ statusCode }) => {
  if (statusCode === 401 || statusCode === 403) return AUTH_ERROR;
  return COMM_ERROR;
};

const initialState = { pendingRequests: 0 };
const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case START_REQUEST:
      return {
        ...state,
        pendingRequests: state.pendingRequests + 1,
      };
    case REQUEST_ERROR:
      return {
        ...state,
        error: { ...action.error, id: pathOr(0, ['error', 'id'], state) + 1, type: getErrorType(action.error) },
        pendingRequests: state.pendingRequests - 1,
      };
    case END_REQUEST:
      return {
        ...state,
        pendingRequests: state.pendingRequests - 1,
      };
    case ACK_ERROR:
      return {
        ...state,
        error: null,
      };
    default:
      return state;
  }
};

export default reducer;
