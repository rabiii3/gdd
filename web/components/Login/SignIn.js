import React from 'react';
import { compose } from 'ramda';
import { connect } from 'react-redux';
import { Form, Field } from 'react-final-form';
import PropTypes from 'prop-types';
import { withRouter, Link } from 'react-router-dom';
import { createStructuredSelector } from 'reselect';
import { bindActionCreators } from 'redux';
import { FormattedMessage } from 'react-intl';
import Button from '@material-ui/core/Button';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import Typography from '@material-ui/core/Typography';
import { withStyles } from '@material-ui/core/styles';
import TextField from '../../forms/TextField';
import withRoutes from '../../hoc/routes';
import messages from './messages';
import { login } from '../../actions/auth';
import { hasSignInError } from '../../selectors/core';
import { required, composeValidators, emailFormat } from '../../forms/utils';

const styles = theme => ({
  button: {
    margin: theme.spacing.unit,
    color: 'white',
    float: 'right',
  },
  link: {
    float: 'right',
    right: theme.spacing.unit,
  },
  textError: {
    color: '#ea4554',
  },
  textField: {
    marginLeft: theme.spacing.unit,
    marginRight: theme.spacing.unit,
    width: 300,
  },
});

export const SignInForm = ({ routes, classes, actions, errorSignIn }) => {
  const onSubmit = values => {
    actions.login(values);
  };

  return (
    <Form
      onSubmit={onSubmit}
      render={({ handleSubmit, pristine, form, invalid }) => (
        <form onSubmit={handleSubmit}>
          <DialogContent>
            {errorSignIn && (
              <Typography variant="body1" gutterBottom align="left" color="secondary">
                <FormattedMessage {...messages.errorSignIn} />
              </Typography>
            )}
            <Field
              validate={composeValidators(
                required(<FormattedMessage {...messages.requiredValidation} />),
                emailFormat(<FormattedMessage {...messages.emailValidation} />),
              )}
              className={classes.textField}
              component={TextField}
              label={<FormattedMessage {...messages.email} />}
              name="email"
              type="email"
            />
            <Field
              validate={composeValidators(required(<FormattedMessage {...messages.requiredValidation} />))}
              className={classes.textField}
              component={TextField}
              label={<FormattedMessage {...messages.password} />}
              name="password"
              type="password"
            />
            <br />
            <br />
            <Button
              component={Link}
              to={routes.getPathByName('requestResetPassword')}
              color="primary"
              align="left"
            >
              {<FormattedMessage {...messages.forgetPW} />}
            </Button>
          </DialogContent>
          <DialogActions>
            <Button
              type="submit"
              disabled={pristine || invalid}
              variant="contained"
              color="primary"
              className={classes.button}
            >
              {<FormattedMessage {...messages.signIn} />}
            </Button>
          </DialogActions>
        </form>
      )}
    />
  );
};

SignInForm.propTypes = {
  classes: PropTypes.object.isRequired,
  actions: PropTypes.object.isRequired,
  errorSignIn: PropTypes.bool,
  routes: PropTypes.object.isRequired,
};

const actions = { login };

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions, dispatch),
});

const mapStateToProps = createStructuredSelector({
  errorSignIn: hasSignInError,
});

export const enhance = compose(
  withStyles(styles),
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  withRouter,
  withRoutes,
);
export default enhance(SignInForm);
