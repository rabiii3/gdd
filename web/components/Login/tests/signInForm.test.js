import React from 'react';
import { shallow } from 'enzyme';
import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import configureStore from '../../../store/configureStore.dev';
import { createBrowserHistory } from 'history';
import { enhance } from '../SignIn';

describe('app | components | signInForm', () => {
  describe('render', () => {
    it('should signInForm render', () => {
      const theme = createMuiTheme({
        typography: {
          useNextVariants: true,
        },
      });
      const Component = () => <div />;
      const EnhancedComponent = enhance(Component);
      const state = { auth: {} };
      const hook = {};
      const store = configureStore(state, hook);
      const history = createBrowserHistory();

      const wrapper = shallow(
        <MuiThemeProvider theme={theme}>
          <Provider store={store}>
            <Router history={history}>
              <EnhancedComponent />
            </Router>
          </Provider>
        </MuiThemeProvider>,
      );

      expect(wrapper).toMatchSnapshot();
    });
  });
});
