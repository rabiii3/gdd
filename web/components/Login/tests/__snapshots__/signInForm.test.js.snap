// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app | components | signInForm render should signInForm render 1`] = `
<Provider
  store={[Function]}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <WithStyles(Connect(withRouter(EnhancedComponent))) />
  </Router>
</Provider>
`;
